File: python/src/threads/evals/live_env.py
"""What the live checks share (spec lane 22, C and lane 32, B): the options a live run is given,
the eval principal, one run of an agent as a value, and reading a live thread's log back."""

import json
from collections.abc import Sequence
from dataclasses import dataclass
from typing import Final, Literal, Protocol

from pydantic import JsonValue

from threads.agents.config import ConfigError
from threads.agents.definition import Definition
from threads.agents.results import Failed, RunResult, StreamEvent, Thread
from threads.agents.run import RunOptions, execute
from threads.agents.store import Store, now_ms, open_store
from threads.evals.report import add_cost
from threads.log import BranchId, Budget, Cost, Event, Principal, ThreadId, UnknownEvent
from threads.loop.guard import ModelBlockedError
from threads.loop.model import Model
from threads.loop.stubs import Stubs
from threads.result import Ok
from threads.store.lines import uuid7
from threads.thread.read import read_log

EVAL_PRINCIPAL: Final = Principal(issuer="threads", tenant="evals", subject="eval-runner")


@dataclass(frozen=True, slots=True)
class Live:
    """What `live` / `--live` gives: the judge model, the budget of every run (the existing
    `Budget`), criteria added after each case's own, and the model that plays a simulated user."""

    judge: Model
    budget: Budget
    rubric: tuple[str, ...] = ()
    user: Model | None = None


class EvalAgent(Protocol):
    """An agent the runner can pin and run: an `agent(...)` whose tools take no deps."""

    @property
    def name(self) -> str: ...

    @property
    def definition(self) -> Definition[None]: ...


@dataclass(frozen=True, slots=True)
class Env:
    live: Live
    store: Store
    kept: bool
    """Thread ids are reported only when the threads are kept (a store the caller passed)."""
    fresh: bool
    """The agent runs on a fresh sandbox: a live eval never restores the case's snapshot."""


@dataclass(frozen=True, slots=True)
class Simulation:
    """EvalCaseResult.simulation (spec lane 32, E): what the conversation did."""

    messages: int
    prefix: Literal["continued", "redriven", "none"]
    prefix_turns: int
    ended: Literal["user_done", "script_done", "max_messages"]
    user_thread_id: str | None = None

    def to_json(self) -> JsonValue:
        out: dict[str, JsonValue] = {
            "messages": self.messages,
            "prefix": self.prefix,
            "prefix_turns": self.prefix_turns,
            "ended": self.ended,
        }
        if self.user_thread_id is not None:
            out["user_thread_id"] = self.user_thread_id
        return out


@dataclass(frozen=True, slots=True)
class Outcome:
    kind: Literal["graded", "error", "skipped", "blocked"]
    check: JsonValue = None
    reason: str = ""
    agent_calls: int = 0
    user_calls: int = 0
    judge_calls: int = 0
    cost: Cost | None = None
    simulation: Simulation | None = None


@dataclass(frozen=True, slots=True)
class Blocked:
    """The model-request guard stopped the run before dispatch."""

    model: str


@dataclass(frozen=True, slots=True)
class Misconfigured:
    """The agent's setup failed: a ConfigError, as its code and message."""

    reason: str


def _drop(_item: StreamEvent) -> None:
    pass


async def events_of(store: Store, branch: BranchId) -> tuple[Event, ...]:
    """A thread's events, read from the store: the log is the truth the report derives from."""
    read = await read_log(store, branch)
    if not isinstance(read, Ok):
        return ()
    return tuple(e for e in read.value.fold.events if not isinstance(e, UnknownEvent))


async def cost_sum(threads: Sequence[Thread]) -> Cost | None:
    """Thread.cost(tree=True) summed; None once any part ran unpriced."""
    total: Cost | None = None
    for i, t in enumerate(threads):
        got = await t.cost(tree=True)
        total = add_cost(total, got.value if isinstance(got, Ok) else None, first=i == 0)
    return total


async def new_thread(env: Env) -> Thread:
    """A new thread for one live run or one simulated conversation."""
    sq = await open_store(env.store)
    now = now_ms()
    thread_id, branch_id = ThreadId(uuid7(now)), BranchId(uuid7(now))
    await sq.create(thread_id, branch_id, now)
    return Thread(thread_id, branch_id, env.store)


async def run_agent(  # noqa: PLR0913, PLR0917 - one run, and what it answers from
    definition: Definition[None],
    text: str,
    env: Env,
    thread: Thread | None,
    budget: Budget | None = None,
    stubs: Stubs | None = None,
) -> RunResult[str] | Blocked | Misconfigured:
    """One run: its result, a guard block, or a setup mistake. Any other raise is a bug. A
    simulated conversation passes one gateway every turn, so its stub queue spans them all."""
    on = thread if thread is not None else await new_thread(env)
    options: RunOptions[None] = {
        "store": env.store,
        "budget": budget if budget is not None else env.live.budget,
        "principal": EVAL_PRINCIPAL,
        "thread": on,
    }
    try:
        return await execute(definition, text, options, None, _drop, stubs=stubs)
    except ModelBlockedError as blocked:
        return Blocked(blocked.model)
    except ConfigError as error:
        return Misconfigured(f"{error.code}: {error.message}")


def answer_of(definition: Definition[None], output: str) -> JsonValue:
    """RunResult.output as the judge sees it: JSON with an output model, else the text."""
    if definition.output is None:
        return output
    parsed: JsonValue = json.loads(output)
    return parsed


def unfinished(result: RunResult[str]) -> str:
    """A run that didn't complete, as the case's error reason."""
    return f"failed: {result.error.code}" if isinstance(result, Failed) else result.status

File: typescript/packages/core/src/evals/live-env.ts
import { z } from "zod";
import { ConfigError } from "../agent/errors";
import type { RunResult, ThreadRef } from "../agent/result";
import { openStore, type Store } from "../agent/sqlite";
import type { Budget, Cost, KnownEvent, Principal } from "../log";
import type { Model } from "../model";
import { ModelBlockedError } from "../model/guard";
import { knownEvents } from "../reduce";
import type { Thread } from "../thread";
import { addCost } from "./report";
import type { JudgeCheck, Simulation } from "./schema";

// What the live checks share (spec lane 22, C and lane 32, B): the options a live run is given,
// the eval principal, the per-case outcome as a value, and reading a live thread's log back.

/** What `live` / `--live` gives: the judge model, the budget of every run, extra criteria. */
export type Live = {
  readonly judge: Model;
  readonly budget: z.infer<typeof Budget>;
  readonly rubric?: readonly string[];
  /** The model that plays the user of a `kind: "model"` simulated case (spec lane 32). */
  readonly user?: Model;
};

export const EVAL_PRINCIPAL: Principal = {
  issuer: "threads",
  tenant: "evals",
  subject: "eval-runner",
};

export type Calls = {
  readonly agent: number;
  readonly user: number;
  readonly judge: number;
};

export const NO_CALLS: Calls = { agent: 0, user: 0, judge: 0 };

export type LiveOutcome =
  | {
      readonly kind: "graded";
      readonly check: JudgeCheck;
      readonly calls: Calls;
      readonly cost: Cost | null;
      readonly simulation?: Simulation;
    }
  | {
      readonly kind: "error" | "skipped";
      readonly reason: string;
      readonly calls: Calls;
      readonly cost: Cost | null;
      readonly simulation?: Simulation;
    }
  | { readonly kind: "blocked"; readonly model: string };

export type Env = {
  readonly live: Live;
  readonly store: Store;
  /** Thread ids are reported only when the threads are kept (a store the caller passed). */
  readonly kept: boolean;
  /**
   * The agent runs on a fresh sandbox: a live eval never restores the case's snapshot, which
   * lives in the provider the case was recorded on.
   */
  readonly fresh: boolean;
};

/** A thread's events, read from the store: the log is the truth the report is derived from. */
export async function eventsOf(
  store: Store,
  thread: ThreadRef,
): Promise<readonly KnownEvent[]> {
  const { log } = await openStore(store);
  const read = await log.read(thread.branch);
  return read.ok ? knownEvents(read.value) : [];
}

/** Thread.cost({tree: true}) summed; null once any part ran unpriced. */
export async function costOf(threads: readonly Thread[]): Promise<Cost | null> {
  let total: Cost | null | undefined;
  for (const t of threads) {
    const got = await t.cost({ tree: true });
    total = addCost(total, got.ok ? got.value : null);
  }
  return total ?? null;
}

/** RunResult.output as the judge sees it: JSON when the agent has an output schema, else text. */
export function answerOf(output: unknown): z.core.util.JSONType {
  return z.json().parse(output ?? null);
}

/** A run that didn't complete, as the case's error reason. */
export function unfinished(
  r: Exclude<RunResult<unknown>, { status: "completed" }>,
): string {
  return r.status === "failed" ? `failed: ${r.error.code}` : r.status;
}

/** A guard block and a setup mistake are values; any other throw is a bug. */
export async function guarded<T>(
  run: () => Promise<T>,
): Promise<T | { readonly blocked: string } | { readonly config: string }> {
  try {
    return await run();
  } catch (error) {
    if (error instanceof ModelBlockedError) return { blocked: error.model };
    if (error instanceof ConfigError)
      return { config: `${error.code}: ${error.message}` };
    throw error;
  }
}
